test/extensions/common/matcher/domain_matcher_test.cc

```diff
@@ -0,0 +1,814 @@
+  NiceMock<Server::Configuration::MockServerFactoryContext> factory_context_;
+  MatchTreeFactory<TestData, absl::string_view> factory_;
+  xds::type::matcher::v3::Matcher matcher_;
+  SkippedMatchCb skipped_match_cb_ = nullptr;
+};
+
+TEST_F(DomainMatcherTest, BasicDomainMatching) {
+  const std::string yaml = R"EOF(
+matcher_tree:
+  input:
+    name: input
+    typed_config:
+      "@type": type.googleapis.com/google.protobuf.StringValue
+  custom_match:
+    name: domain_matcher
+    typed_config:
+      "@type": type.googleapis.com/xds.type.matcher.v3.ServerNameMatcher
+      domain_matchers:
+      - domains:
+        - "api.example.com"
+        on_match:
+          action:
+            name: test_action
+            typed_config:
+              "@type": type.googleapis.com/google.protobuf.StringValue
+              value: exact_match
+      - domains:
+        - "*.example.com"
+        on_match:
+          action:
+            name: test_action
+            typed_config:
+              "@type": type.googleapis.com/google.protobuf.StringValue
+              value: wildcard_match
+      - domains:
+        - "*.com"
+        on_match:
+          action:
+            name: test_action
+            typed_config:
+              "@type": type.googleapis.com/google.protobuf.StringValue
+              value: global_wildcard
+  )EOF";
+  loadConfig(yaml);
+
+  // Test exact match (highest priority)
+  {
+    auto input = TestDataInputStringFactory("api.example.com");
+    validateMatch("exact_match");
+  }
+
+  // Test wildcard match (middle priority)
+  {
+    auto input = TestDataInputStringFactory("foo.example.com");
+    validateMatch("wildcard_match");
+  }
+
+  // Test broader wildcard match (lower priority)
+  {
+    auto input = TestDataInputStringFactory("something.else.com");
+    validateMatch("global_wildcard");
+  }
+}
+
+TEST_F(DomainMatcherTest, WildcardPrecedenceOrder) {
+  const std::string yaml = R"EOF(
+matcher_tree:
+  input:
+    name: input
+    typed_config:
+      "@type": type.googleapis.com/google.protobuf.StringValue
+  custom_match:
+    name: domain_matcher
+    typed_config:
+      "@type": type.googleapis.com/xds.type.matcher.v3.ServerNameMatcher
+      domain_matchers:
+      - domains:
+        - "*.com"
+        on_match:
+          action:
+            name: test_action
+            typed_config:
+              "@type": type.googleapis.com/google.protobuf.StringValue
+              value: short_wildcard
+      - domains:
+        - "*.example.com"
+        on_match:
+          action:
+            name: test_action
+            typed_config:
+              "@type": type.googleapis.com/google.protobuf.StringValue
+              value: long_wildcard
+  )EOF";
+  loadConfig(yaml);
+
+  // Should match the longer suffix wildcard (*.example.com) first
+  // Declaration order matters when multiple matchers are present
+  {
+    auto input = TestDataInputStringFactory("api.example.com");
```

Comment:
This does not quite make sense to me. The longest suffix should win, right?

Reply:
Yeah, I incorrectly assumed declaration order mattered for different specificity levels. I've fixed the test to verify that longest suffix wins regardless of declaration order.